### PR TITLE
devices/raw-flash: documentation for raw flash support

### PR DESCRIPTION
**Pull request by bboozzoo**
@drewmoseley @estenberg @kacf

### CODE REVIEW

**Comment by bboozzoo** [comment]
review anyone?

**Comment by estenberg** [comment]
I can only give structure/surface-level review on this, @drewmoseley should cover the deeper technical aspects about the tutorial. A few notes from me:
* The section at https://docs.mender.io/1.1/devices/partition-layout#flash-memory-types needs to be updated to link to raw flash, and remove the note that Mender does not support raw flash
* Also review the rest of https://docs.mender.io/1.1/devices/partition-layout if it needs adjustment, links or integration with the new raw flash doc
* U-Boot is spelled with capital U and B, at least that's what we do in the rest of the docs.

**Comment by drewmoseley** [comment]
Indeed the canonical spelling per the Denx web-page is U-Boot.  We should definitely stick to that.

@estenberg you specifical mentioned the links to the 1.1 docs.  Is your expectation for this to go to 1.1 or just to development?

File: 03.Devices/05.Raw-flash/docs.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
+---
+title: Raw flash
+taxonomy:
+    category: docs
+---
+
+This chapter introduces technical details of raw flash support in Mender.
+
+As noted [partition layout](../partition-layout) chapter, Mender officially
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+supports *block devices* and there is *experimental* support for raw flash and
+UBI volumes and file systems.
+
+## Overview
+
+Support for raw flash memory under Linux, is in general more complicated than
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: done

File: 03.Devices/05.Raw-flash/docs.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+kernel and rootfs before introducing Mender.
+
+As an example, to illustrate potential pain points we will use a Versatile
+Express CortexA9x4 board, emulated under qemu (`vexpress-a9` target). The board
+comes with 128MB of CFI NOR flash, provided in form of 2 * 64MB dies. Respective
+details may slightly differ for NAND flash or SPI NOR flash devices.
+
+### Mender
+
+Mender Yocto layer comes with support for the following raw flash boards in the
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: done

File: 03.Devices/05.Raw-flash/docs.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+main tree:
+
+* [Toradex Colibri VF50/VF61 `meta-mender-toradex-nxp`](https://github.com/mendersoftware/meta-mender/tree/master/meta-mender-toradex-nxp)
+* [vexpress-a9 emulated qemu `meta-mender-qemu`](https://github.com/mendersoftware/meta-mender/tree/master/meta-mender-qemu)
+
+Both layers provide a set of patches enabling for their upstream u-boots and can
+be used as reference.
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: replaced

File: 03.Devices/05.Raw-flash/docs.md [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+### qemu
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: replaced qemu with QEMU where relevant

File: 03.Devices/05.Raw-flash/docs.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+It is possible to build a Yocto image and inspect all the details of Mender
+integration by adding `meta-mender-qemu` layer to the build. The layer defines a
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+`vexpress-qemu-flash` machine and includes all necessary pieces to enable MTD
+and UBI support.
+
+!! In order to test raw flash support under qemu, a qemu version >= 2.9 is required. Earlier versions contain a bug in CFI flash support that renders flash support on `vexpress-a9` unusable.
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+```bash
+INHERIT += "mender-full-ubi"
+
+...
+
+MACHINE = "vexpress-qemu-flash"
+
+```
+
+Successful build will produce a `vexpress-nor` image in `${DEPLOYDIR}`:
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+```bash
+$ ls -shLl tmp/deploy/images/vexpress-qemu-flash/core-image-minimal-vexpress-qemu-flash.vexpress-nor
+129M -rw-r--r-- 2 user user 129M 07-18 15:34 tmp/deploy/images/vexpress-qemu-flash/core-image-minimal-vexpress-qemu-flash.vexpress-nor
+```
+
+A `vexpress-nor` image is a tar file that contains an image for each of `nor`
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+'drives' emulated by qemu.
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+
+The image can be run by calling a `mender-qemu` helper script provided in
+`meta-mender-qemu` layer:
+
+```bash
+QEMU_SYSTEM_ARM=$HOME/qemu-install/bin/qemu-system-arm \
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: added a note

File: 03.Devices/05.Raw-flash/docs.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+VEXPRESS_IMG=tmp/deploy/images/vexpress-qemu-flash/core-image-minimal-vexpress-qemu-flash.vexpress-nor \
+MACHINE=vexpress-qemu-flash \
+    ../meta-mender/meta-mender-qemu/scripts/mender-qemu
+```
+
+### Raw flash devices, partitioning
+
+Using raw flash devices under Linux is more complicated compared to typical
+block devices such as hard disks, or eMMC flash. Block devices typically come
+with a partition table, either using MBR or GPT, that enables discovery and
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+QEMU_SYSTEM_ARM=$HOME/qemu-install/bin/qemu-system-arm \
+VEXPRESS_IMG=tmp/deploy/images/vexpress-qemu-flash/core-image-minimal-vexpress-qemu-flash.vexpress-nor \
+MACHINE=vexpress-qemu-flash \
+    ../meta-mender/meta-mender-qemu/scripts/mender-qemu
+```
+
+### Raw flash devices, partitioning
+
+Using raw flash devices under Linux is more complicated compared to typical
+block devices such as hard disks, or eMMC flash. Block devices typically come
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+with a partition table, either using MBR or GPT, that enables discovery and
+identification of existing partitions. In case of raw flash devices, no
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+partition tables are in use. Instead, the kernel must be informed about existing
+partitions, their start locations and sizes. This can achieved using a number of
+different methods:
+
+* machine configuration via `arch/<ARCH>/mach-<MACH>` (deprecated)
+* kernel command line ([drivers/mtd/cmdlinepart.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/mtd/cmdlinepart.c)) typically set up by bootloader
+* device tree ([DTS MTD bindings](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/plain/Documentation/devicetree/bindings/mtd/partition.txt))
+
+Raw flash boards currently supported by Mender use kernel command line to pass
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+information about MTD partitions.
+
+Using `vexpress-a9` qemu target as an example, the flash area is partitioned like this:
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+
+* 1MB - u-boot, partition name `u-boot`
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+* 1MB - u-boot environment (main + redundant), parition name `u-boot-env`
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+* remaining space (126MB) - UBI, parition name `ubi`
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+
+Having enabled `CONFIG_CMD_MTDPARTS` in u-boot we can see the following output
+after issuing `mtdparts` at u-boot prompt:
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+ 2: ubi                 0x07e00000      0x00200000      0
+
+active partition: nor2,0 - (u-boot) 0x00100000 @ 0x00000000
+
+defaults:
+mtdids  : nor2=40000000.flash
+mtdparts: mtdparts=40000000.flash:1m(u-boot)ro,1m(u-boot-env),-(ubi)
+```
+
+Note that `mtdparts` command line argument is using the same device name as produced by board devicetree bindings.
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+root@vexpress-qemu-flash:~# cat /proc/mtd 
+dev:    size   erasesize  name
+mtd0: 00100000 00040000 "u-boot"
+mtd1: 00100000 00040000 "u-boot-env"
+mtd2: 07e00000 00040000 "ubi"
+```
+
+### UBI and UBI File System
+
+`ubinize` and `mkfs.ubifs` arguments are a little complicated to get right. One
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+can use `mtdinfo` in a running system to obtain a set a reasonable defaults
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: added

File: 03.Devices/05.Raw-flash/docs.md [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+(using `vexpress-a9` qemu target as an example):
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: replaced

File: 03.Devices/05.Raw-flash/docs.md [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+Character device major/minor:   90:4
+Bad blocks are allowed:         false
+Device is writable:             true
+Default UBI VID header offset:  64
+Default UBI data offset:        128
+Default UBI LEB size:           262016 bytes, 255.9 KiB
+Maximum UBI volumes count:      128
+```
+
+Note that these settings will generally be different depending on type of flash
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+memory.
+
+Once determined, the paramters for `mkfs.ubifs` and `ubinize` must be set in
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+Yocto configuration using `MKUBIFS_ARGS` and `UBINIZE_ARGS` variables
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 28]
```diff
@@ -0,0 +1,290 @@
...
+respectively. Since, these paratmeters are a specific for given board, it is
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 29]
```diff
@@ -0,0 +1,290 @@
...
+possible they may already be set by a corresponding machine configuration.
+
+## u-boot
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 30]
```diff
@@ -0,0 +1,290 @@
...
+
+To enable UBI support in u-boot and integrate it with the kernel you will need
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 31]
```diff
@@ -0,0 +1,290 @@
...
+	"ubifsmount ubi0:rootfs && "                              \
+	"ubifsload ${kernel_addr_r} /boot/zImage && "             \
+	"ubifsload ${fdt_addr_r} /boot/${fdtfile} && "            \
+    "setenv bootargs ${mtdparts} ${ubiargs} ${defargs} && "   \
+	"bootz ${kernel_addr_r} - ${fdt_addr_r}\0"
+```
+
+The script:
+
+1. attacheds UBI partition using `ubi` devince defined by `mtdparts`, creating
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: AFAIK attaching is the right term here, see this u-boot log for reference:

```Booting from NAND...
ubi0: default fastmap pool size: 200
ubi0: default fastmap WL pool size: 100
ubi0: attaching mtd1
ubi0: scanning is finished
ubi0: attached mtd1 (name "mtd=3", size 509 MiB)
```

File: 03.Devices/05.Raw-flash/docs.md [hunk 32]
```diff
@@ -0,0 +1,290 @@
...
+   `ubi0` device
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 33]
```diff
@@ -0,0 +1,290 @@
...
+2. mounts UBIFS volume `rootfs` from `ubi0`
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 34]
```diff
@@ -0,0 +1,290 @@
...
+3. loads `zImage` at kernel load address 0x60100000 (start of RAM + 1MB offset)
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 35]
```diff
@@ -0,0 +1,290 @@
...
+4. loads flattened device tree at 0x60000000 (start of ram)
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 36]
```diff
@@ -0,0 +1,290 @@
...
+5. updates `bootargs` to include:
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 37]
```diff
@@ -0,0 +1,290 @@
...
+   1. `mtdparts` - MTD partitions, locations, sizes and naming
+   2. `ubiargs` - MTD device carrying UBI, root file system location contents on `ubi0:rootfs` volume and sets rootfs type to UBIFS
+   3. `defargs` - additional default arguments, such as console, panic settings and similar
+6. loads the kernel
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 38]
```diff
@@ -0,0 +1,290 @@
...
+* `MENDER_PARTITION_ALIGNMENT_KB` - partition alignment, set to match erase block size
+
+When using UBI you may need to set `MENDER_STORAGE_RESERVED_RAW_SPACE` to
+account for space lost to UBI metadata. 
+
+These settings affect the calculated rootfs size.
+
+!!! Note that the calculated rootfs size (i.e. volume size) is different from the actual amount of data that can be stored in rootfs. The difference is caused by compression. To set an upper boundary on the amount of rootfs data, you can define `IMAGE_ROOTFS_MAXSIZE`.
+
+### u-boot integration
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 39]
```diff
@@ -0,0 +1,290 @@
...
+`mender-install.bbclass`. 
+
+Mender support will create a UBI image file (`ubimg` in `${DEPLOYDIR}/images`)
+including the following volumes:
+
+* `rootfsa` - contents of root filesystem A
+* `rootfsb` - contents of root filesystem B
+* `data` - contents of data partition
+
+The `ubimg` image file can be used for populating UBI partition with `ubiformat`
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 40]
```diff
@@ -0,0 +1,290 @@
...
+When using UBI you may need to set `MENDER_STORAGE_RESERVED_RAW_SPACE` to
+account for space lost to UBI metadata. 
+
+These settings affect the calculated rootfs size.
+
+!!! Note that the calculated rootfs size (i.e. volume size) is different from the actual amount of data that can be stored in rootfs. The difference is caused by compression. To set an upper boundary on the amount of rootfs data, you can define `IMAGE_ROOTFS_MAXSIZE`.
+
+### u-boot integration
+
+For u-boot, on top of options listed in [u-boot](#u-boot) you will need to
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 41]
```diff
@@ -0,0 +1,290 @@
...
+enable options required by Mender listed
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 42]
```diff
@@ -0,0 +1,290 @@
...
+in [u-boot integration](../integrating-with-u-boot#u-boot-features).
+
+The u-boot boot process remains very much the same as described
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 43]
```diff
@@ -0,0 +1,290 @@
...
+in [integration points](../integrating-with-u-boot#integrtion-points) with
+addition of a call to `mender_setup` script (using `vexpress-a9` as an example):
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 44]
```diff
@@ -0,0 +1,290 @@
...
+"ubiboot=" \
+	"echo Booting from NOR...; "                              \
+	"run mender_setup; "                                      \
+	"run set_ubiargs; "                                       \
+	"ubi part ${mender_mtd_ubi_dev_name} && "                 \
+	"ubifsload ${kernel_addr_r} /boot/zImage && "             \
+    ...
+```
+
+Note that u-boot places some constraints on parameter expansion, for this reason
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 45]
```diff
@@ -0,0 +1,290 @@
...
+the paramter `ubiargs` is no longer set by default environment, but instead it
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: fixed

File: 03.Devices/05.Raw-flash/docs.md [hunk 46]
```diff
@@ -0,0 +1,290 @@
...
+is set by intermediate `set_ubiargs` script.
```
Comment (bboozzoo) [inline 03.Devices/05.Raw-flash/docs.md]: and fixed

**Comment by bboozzoo** [comment]
@drewmoseley thanks for proof reading. I've added a note about checking the output of `qemu-system-arm --version` command.

**Comment by bboozzoo** [comment]
@eysteins are you saying that we should remove a note about raw flash support being experimental?

**Comment by estenberg** [comment]
@drewmoseley sorry about the confusion with the 1.1 docs links - we should not touch 1.1, the raw flash support is in 1.2 (i.e. master today).
@bboozzoo yes, from 1.2 raw flash will not be experimental, it will be supported. So please remove the note about raw flash being experimental in master docs (which becomes 1.2).

**Comment by drewmoseley** [comment]
Regarding the requirement for an updated qemu version, shouldn't we include the necessary fixes in our layer so that the standard one built by bitbake will be usable?  Relying on external tools is generally avoided when using Yocto.  @estenberg @kacf thoughts?

File: 03.Devices/02.Partition-layout/docs.md [hunk 47]
```diff
@@ -47,14 +47,14 @@ reason, they are also sometimes referred to as **Flash Translation Layer (FTL) d
 For example, these types of flash devices expose themselves as block devices: **SD, mini-SD, micro-SD,
 MMC, eMMC, RS-MMC, SSD, USB, CompactFlash, MemoryStick, MemoryStick Micro**.
 
-* **Raw flash.** Raw flash devices do not have a memory controller
-that takes care of wear leveling nor error correction, so this *must be handled in software*.
-In Linux, raw flash devices are exposed as a **Memory Technology Device (MTD)** file.
-Care must be taken when selecting a file system that is MTD-aware, since
-it should handle the wear leveling and error correction.
+* **Raw flash.** Raw flash devices do not have a memory controller that takes
+care of wear leveling nor error correction, so this *must be handled in
```
Comment (drewmoseley) [inline 03.Devices/02.Partition-layout/docs.md]: wear leveling _and_ error correction.

File: 03.Devices/02.Partition-layout/docs.md [hunk 48]
```diff
@@ -47,14 +47,14 @@ reason, they are also sometimes referred to as **Flash Translation Layer (FTL) d
...
+software*. In Linux, raw flash devices are exposed as a **Memory Technology
+Device (MTD)** file. Care must be taken when selecting a file system that is
+MTD-aware, since it should handle the wear leveling and error correction.
```
Comment (drewmoseley) [inline 03.Devices/02.Partition-layout/docs.md]: Care must be taken when selecting a file system to ensure that it is MTD-aware and properly handles wear leveling and error correction.

File: 03.Devices/02.Partition-layout/docs.md [hunk 49]
```diff
@@ -47,14 +47,14 @@ reason, they are also sometimes referred to as **Flash Translation Layer (FTL) d
...
 Popular file systems for MTD devices include UBIFS, JFFS2, and YAFFS.
-
-! Mender officially supports  *block devices* and there is *experimental* support for raw flash and UBI volumes and file systems. To enable UBI support, inherit the `mender-full-ubi` class in your `local.conf` and take a look at the various UBI related variables in `mender-install.bbclass`. Please keep in mind that this is an experimental feature so please provide feedback on the [mailing list](https://groups.google.com/a/lists.mender.io/forum?target=_blank#!forum/mender) if you have any questions or issues. Raw flash support will be officially supported once it has been thoroughly tested and it has proven high quality.
+Consult [raw flash](../raw-flash) section for details on setting up and
```
Comment (drewmoseley) [inline 03.Devices/02.Partition-layout/docs.md]: Consult _the_ raw flash section...

File: 03.Devices/05.Raw-flash/docs.md [hunk 50]
```diff
@@ -0,0 +1,290 @@
+---
+title: Raw flash
+taxonomy:
+    category: docs
+---
+
+This chapter introduces technical details of raw flash support in Mender.
```
Comment (drewmoseley) [inline 03.Devices/05.Raw-flash/docs.md]: introduces _the_ technical ...

File: 03.Devices/05.Raw-flash/docs.md [hunk 51]
```diff
@@ -0,0 +1,290 @@
...
+including the following volumes:
+
+* `rootfsa` - contents of root filesystem A
+* `rootfsb` - contents of root filesystem B
+* `data` - contents of data partition
+
+The `ubimg` image file can be used for populating the UBI partition with
+the `ubiformat` tool.
+
+By default a `*.ubifs` root filesystem image is used as a Mender artifact.
```
Comment (drewmoseley) [inline 03.Devices/05.Raw-flash/docs.md]: I am not sure what this is saying.  I think it should be
By default a \`*.ubifs\` root filesystem image will be used when generating a Mender artifact.

File: 03.Devices/05.Raw-flash/docs.md [hunk 52]
```diff
@@ -0,0 +1,290 @@
...
+	"echo Booting from NOR...; "                              \
+	"run mender_setup; "                                      \
+	"run set_ubiargs; "                                       \
+	"ubi part ${mender_mtd_ubi_dev_name} && "                 \
+	"ubifsload ${kernel_addr_r} /boot/zImage && "             \
+    ...
+```
+
+Note that U-Boot places some constraints on parameter expansion, for this reason
+the paramter `ubiargs` is no longer set by default environment. Instead it is
```
Comment (drewmoseley) [inline 03.Devices/05.Raw-flash/docs.md]: parameter is misspelled.

**Comment by bboozzoo** [comment]
ping

**Comment by drewmoseley** [comment]
Looks good.  I'd like to wait for review from @kacf when he returns before merging.

**Comment by bboozzoo** [comment]
Integration checklist updated where necessary.

I'm not sure about:
```
stat -c %D /
stat -c %t%02T /dev/mmcblk0p2
```
Don't think this will work with UBIFS (or device mapper & LVM for that matter).

File: 03.Devices/04.Integration-checklist/docs.md [hunk 53]
```diff
@@ -128,36 +133,56 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
 
 16. Repeat step 12, but this time verify that the build date is later than the first time you did this step. This is because we replaced the kernel of the second rootfs partition with a newer one. This verifies that the correct kernel is loaded from partition B.
 
-17. Repeat step 13, but this time using `/dev/mmcblk0p3`. This verifies that the correct rootfs is mounted when partition B is active.
+    !!! In the case of raw flash storage, the second rootfs was not updated, hence the build time listed in `/proc/version` will be the same as for the first partition.
+
+17. Repeat step 13, but this time using `/dev/mmcblk0p3` (or `ubi0_1`). This verifies that the correct rootfs is mounted when partition B is active.
 
-    ! As in the previous step, `/dev/mmcblk0p3` may be different if `MENDER_ROOTFS_PART_B` or `MENDER_STORAGE_DEVICE_BASE` have been changed.
+    ! As in the previous step, `/dev/mmcblk0p3` (or `ubi0_1`) may be different if `MENDER_ROOTFS_PART_B` or `MENDER_STORAGE_DEVICE_BASE` have been changed.
```
Comment (drewmoseley) [inline 03.Devices/04.Integration-checklist/docs.md]: has been changed

Comment (bboozzoo) [inline 03.Devices/04.Integration-checklist/docs.md]: fixed

**Comment by kacf** [comment]
> Regarding the requirement for an updated qemu version, shouldn't we include the necessary fixes in our layer so that the standard one built by bitbake will be usable? Relying on external tools is generally avoided when using Yocto.

I think the necessary patches were upstreamed, were they not, @bboozzoo? Poky is moving pretty fast, so I expect this to show up on their branch pretty soon.

Of course the user experience is better if we provide it now, especially for pyro, but it's some work, might clash with the upstream version, and will eventually resolve itself, so not high priority IMHO.

**Comment by bboozzoo** [comment]
merge?

**Comment by kacf** [comment]
I'll let @drewmoseley have the final word!

**Pull request merged**